### DISCUSSION TITLE
Use PorousFlowOutflowBC outside of Porous Flow module architecture

### DISCUSSION THREAD

**Discussion by alexguev** [idaholab/moose#20590] [Q&A General] [answered]
Hi,

I am trying to simulate a Darcy-Cahn-Hilliard model in a Hele-Shaw setting, following the equations (10) and (11) from [this paper](https://www.sciencedirect.com/science/article/pii/S0021999117304527?via%3Dihub). I have implemented them without using the Porous Flow module as I was thinking that adding the coupling with Cahn-Hilliard might not be straightforward (I'll share my input file if that helps). In particular, coupling the initial Buckley–Leverett problem with Cahn-Hilliard provides sufficient regularization, to some extent, to avoid resorting to artificial numerical diffusion or viscosity, while keeping a fairly sharp saturation front. 
In addition, I have poromechanics coupling. In all, I am solving for the saturation, the pressure, the Cahn-Hilliard potential (usual splitting into two second-order PDEs) and the displacement. Note that there is only one pressure as I neglect capillary effects (Buckley–Leverett setting). The porosity is coupled with the volumetric strain and pressure, as per usual poromechanics.

I am interested in simulating the invasion of water into oil for instance. Notably, I have a Neumann BC in pressure at entry to have constant flow rate, and a Dirichlet BC in pressure at exit equal to 0; I want the fluids to be freely expelled out of the domain. However, for certain sets of parameters (e.g. low Young's modulus), the exit boundary can be highly deformed leading to convergence issues; this happens at first timestep and then the boundary relaxes back to a vertical position (see picture below at first timestep). In the setting below, saturation phase-field is shown, injection is on the left boundary, displacement is fixed to 0 everywhere but on exit right boundary (if I fix displacement on exit boundary as well, the mesh will then be distorted).

<img width="621" alt="deformed-bar" src="https://user-images.githubusercontent.com/38299791/159040437-7bb424c2-5ef3-401e-ad70-9b6ec7a535f9.png">

I understand that this problem should be addressed by PorousFlowOutflowBC, to avoid the exit boundary to try and expel fluid that is not there yet, which I think leads to the deformation due to the poromechanics coupling. However, given the architecture of Porous Flow module, I can't really use it outside as such. I thus find myself at a crossroads between:

- trying to implement this type of BC outside of the Porous Flow module architecture (how easy would that be?) and
- implementing my model in the Porous Flow module, which benefits from all the hard work put here to address all that kind of tricky issues related to poromechanics, but might be tedious to implement my equations; in particular, adding the Cahn-Hilliard coupling and modify the existing porous flow equations accordingly. In addition, I will add in the near future further customization regarding the mechanics, which is not in Porous Flow module.

Thanks in advance,

Alex

**Comment by WilkAndy** [comment, marked as answer]
Hi @alexguev , i wrote about half of the PorousFlow module including `PorousFlowOutflowBC`.  Are you a little experienced with MOOSE C++?  If so, i suggest that you write your own version of the BC.  One of the benefits of MOOSE is that it is fairly easy to customize to your own physics, and i suspect your case is a situation where that is optimal.  One advantage here is that you can somewhat copy the code of `PorousFlowOutflowBC`.  You'll be able to omit a lot of that code because you probably don't have multicomponent fluids, and might not need upwinding.

One nice thing about PorousFlow is that it plays nicely with `tensor_mechanics` or `geochemistry`, which are the most common use-cases of inter-module coupling, but, you're right, it is not designed to couple naturally with all parts of MOOSE

**Reply by WilkAndy** [reply]
Just wanted to say: it's really cool that you're doing this work.  Please keep us advised on progress (maybe in this dicussion?).   Also, the above it just my opinion based on only a little thought - no doubt others will have other opinions.

**Reply by alexguev** [reply]
Hey @WilkAndy, thanks for your reply and all the work you've been putting in for this module!
Yes I have just written my own version of the BC mimicking HeatConductionOutflow.C, which doesn't seem to solve my issue in the end (namely the exit boundary still deforms) - although I may have done something wrong.

Since there seems to be many sources of error for such highly nonlinear problem (poromechanical Darcy-Cahn-Hilliard), would there be an example/test that would be fairly similar in the Porous Flow module? So I can see what issues/solutions I should be aware of. To put it briefly, I'm concerned with injecting water into oil in the "beam" domain above (ultimately in a Hele-Shaw cell), with constant flow rate (so Neumann BC in pressure at entry boundary). The oil pushed by water is free to flow out at the exit boundary. In addition, I want to capture the deformation due to pore pressure. Tentatively, I might be able to insert the Cahn-Hilliard bits into such example to benefit from the Porous Flow module framework.

First of, I will check in more details the Convergence criteria page, which seems important in such hydromechanical coupling.

I will be happy to share in more details once I make sure I don't have any too obvious mistakes. It is indeed a very interesting topic!

**Reply by WilkAndy** [reply]
There's not really an example that you can exactly follow.  I agree that the outflow end shouldn't deform too much, since fluid should just flow out!  Similar to your example, there are the following (off the top of my head)

- Buckley-Leverett https://mooseframework.inl.gov/modules/porous_flow/tests/buckley_leverett/buckley_leverett_tests.html
- Cold CO2​ injection into an elastic reservoir - a multi-phase THM problem https://mooseframework.inl.gov/modules/porous_flow/thm_example.html
- Some of the examples towards the bottom of: https://mooseframework.inl.gov/modules/porous_flow/tests/sinks/sinks_tests.html

a
